### PR TITLE
[var] Implement hb_font_get_var_coords_design

### LINKED ISSUE #1865: Recover design coords from a hb_font_t
I want to create a ui for font variations.

I have a hb_font_t. From that I can get the hb_face_t, and all the axis infos to set up sliders.

But there is currently no way for me to find the coords of the hb_font_t to set the sliders current value. hb_font_get_var_coords_normalized gives me _something_, but it does not obviously relate to the min/max/default values in the axis info.

### PR DESCRIPTION
**Pull request by ebraminio**
Hold design coords and simulate when normalized coords are set directly.

Fixes #1865

### CODE REVIEW

File: test/api/test-var-coords.c [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+#include <hb-ot.h>
+
+/* Unit tests for hb_font_[gs]et_var_coords_ */
+
+#ifndef g_assert_cmpfloat_with_epsilon
+#define g_assert_cmpfloat_with_epsilon(n1, n2, epsilon) g_assert_cmpfloat ((int) roundf (n1), ==, (int) roundf (n2))
+#endif
+
+#define EPSILON 0.05
```
Comment (ebraminio) [inline test/api/test-var-coords.c]: Converting back precision, client can round with such precision for getting a reliable design coord

File: test/api/test-var-coords.c [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+  g_assert_cmpint ((int) hb_font_get_var_coords_design (font, NULL)[0], ==, 206);
+
+  for (float weight = 200; weight < 901; ++weight)
+  {
+    int normalized;
+    hb_ot_var_normalize_coords (face, 1, &weight, &normalized);
+    hb_font_set_var_coords_normalized (font, &normalized, 1);
+    float converted_back = hb_font_get_var_coords_design (font, NULL)[0];
+    // fprintf (stderr, "%f: %d => %f\n", weight, normalized, converted_back);
+    g_assert_cmpfloat_with_epsilon ((long double) converted_back, (long double) weight, EPSILON);
```
Comment (ebraminio) [inline test/api/test-var-coords.c]: glib uses long double and if we don't cast we will double promotion.

**Comment by ebraminio** [comment]
As it is now test covered also, PTAL. It unblocks my ot-style work also.

**Comment by ebraminio** [comment]
Some feedback here will be nice as it blocks ot-style and font slant angle retrieval, unless, we decide to expose those in hb_face_t not hb_font_t.

**Comment by behdad** [comment]
Will check soon.  Reviewed the avar/fvar changes.  Is this dependent on that or not?

**Comment by ebraminio** [comment]
> Will check soon. Reviewed the avar/fvar changes. Is this dependent on that or not?

Merged the two.

File: src/hb-font.cc [hunk 3]
```diff
@@ -1389,15 +1407,23 @@ hb_font_create_sub_font (hb_font_t *parent)
...
+  if (num_coords)
   {
-    unsigned int size = parent->num_coords * sizeof (parent->coords[0]);
-    font->coords = (int *) malloc (size);
-    if (unlikely (!font->coords))
-      font->num_coords = 0;
-    else
-      memcpy (font->coords, parent->coords, size);
+    unsigned int coords_size = num_coords * sizeof (parent->coords[0]);
+    int *coords = (int *) malloc (coords_size);
```
Comment (behdad) [inline src/hb-font.cc]: calloc.

File: src/hb-font.cc [hunk 4]
```diff
@@ -1389,15 +1407,23 @@ hb_font_create_sub_font (hb_font_t *parent)
...
+    if (likely (coords))
+    {
+      unsigned int design_coords_size = num_coords * sizeof (parent->design_coords[0]);
+      float *design_coords = (float *) malloc (design_coords_size);
```
Comment (behdad) [inline src/hb-font.cc]: calloc.

Comment (behdad) [reply src/hb-font.cc]: Also, do both calloc calls then check both results together?

File: src/hb-font.cc [hunk 5]
```diff
@@ -1854,10 +1870,25 @@ hb_font_set_variations (hb_font_t *font,
   if (unlikely (coords_length && !normalized))
     return;
 
+  float *design_coords = coords_length ? (float *) calloc (coords_length, sizeof (float)) : nullptr;
+  if (unlikely (coords_length && !design_coords))
+  {
+    free (normalized);
+    return;
+  }
```
Comment (behdad) [inline src/hb-font.cc]: Same here.  Combine the allocation-failure checking.

File: src/hb-font.cc [hunk 6]
```diff
@@ -1854,10 +1870,25 @@ hb_font_set_variations (hb_font_t *font,
...
+
+  for (unsigned int i = 0; i < variations_length; i++)
+  {
+    hb_ot_var_axis_info_t info;
+    if (hb_ot_var_find_axis_info (font->face, variations[i].tag, &info) &&
+	info.axis_index < coords_length)
+      design_coords[info.axis_index] = variations[i].value;
+  }
```
Comment (behdad) [inline src/hb-font.cc]: Rewrite this function to populate design coords completely, then normalize them...

File: src/hb-font.cc [hunk 7]
```diff
@@ -1877,8 +1908,18 @@ hb_font_set_var_coords_design (hb_font_t *font,
   if (unlikely (coords_length && !normalized))
     return;
 
+  float *design_coords = coords_length ? (float *) calloc (coords_length, sizeof (float)) : nullptr;
+  if (unlikely (coords_length && !design_coords))
+  {
+    free (design_coords);
+    return;
+  }
```
Comment (behdad) [inline src/hb-font.cc]: Combine allocation/failure checks.

File: src/hb-font.cc (hunk 5 above)
Comment (ebraminio) [reply src/hb-font.cc]: Done

File: src/hb-font.cc (hunk 4 above)
Comment (ebraminio) [reply src/hb-font.cc]: Done

File: src/hb-font.cc (hunk 3 above)
Comment (ebraminio) [reply src/hb-font.cc]: Done

File: src/hb-font.cc (hunk 7 above)
Comment (ebraminio) [reply src/hb-font.cc]: Done

**Comment by ebraminio** [comment]
We should either have this or make ot-style to get hb_face_t, so

> How are we going to communicate, when avar2 is adopted, that design coords are not available?

There should be some, even non unique, design coord for a normalized coord even after avar2? If so why we should care about uniqueness I wonder

> This is completely lacking in named-instance handling. Might be slightly orthogonal, but I will feel a lot more comfortable if we answer questions in #1883 first.

I see them differently but understand your concern, gave my 2c there.

File: src/hb-font.cc (hunk 6 above)
Comment (ebraminio) [reply src/hb-font.cc]: don't know what "completely" means here but have merged normalize to the code no extra axis fetching.

**Comment by ebraminio** [comment]
Wished not being able for transforming back values for not yet developed version of a not very essential table wasn't blocking things like returning italic angle and other nice things ot-style returns for non-VF fonts but apparently that is the case.

**Comment by ebraminio** [comment]
My opinion, I think we should just announce [not necessarily now and not necessarily going to be removed] `hb_font_set_var_coords_normalized` (thus other normalized coords) as lightly deprecated, why other libraries should know about such internal details of a font, this will be only useful for hb-ft integration and we could do that without public APIs anyway.

File: src/hb-font.cc (hunk 6 above)
Comment (ebraminio) [reply src/hb-font.cc]: Too much of duplication as it used to call `hb_ot_var_normalize_variations` which does another `hb_ot_var_find_axis_info` and if I just normalize every axis it makes `tests/GPOS-5.tests` fail, can upload on a separate PR.

**Comment by ebraminio** [comment]
> How are we going to communicate, when avar2 is adopted, that design coords are not available?

I think let's deal with that when `avar2` become available, let's slightly deprecate APIs dealing with normalized coords I would say as clients should only care about design coords.

> This is completely lacking in named-instance handling. Might be slightly orthogonal, but I will feel a lot more comfortable if we answer questions in #1883 first.

I think this has more importance than that, named instances are also themselves stored as design coords, not normalized coords, and I think I can reach to that sometime after this.

**Comment by ebraminio** [comment]
Okay, lets merge this and work on ways to adopt #1883 also

**Pull request merged**